grammar.js

```diff
@@ -560,11 +565,33 @@ module.exports = grammar({
 
     finally_clause: $ => prec.right(seq('finally', $.expression)),
 
+    binding: $ => prec.dynamic(PREC.binding, seq(
+        field('name', $.identifier),
+        optional(seq(':', field('type', $._param_type))),
+    )),
+
+    bindings: $ => seq(
+        '(',
+        commaSep($.binding),
+        ')',
+    ),
+
+    lambda_expression: $ => prec.right(PREC.lambda, seq(
+        choice(
+            $.bindings,
+            $.identifier,
```

Comment:
why does it have to be a choice of `binding` or `identifier` if `binding` has a `name: (identifier)`?

Reply:
it's a choice of `bindings` (mind the `s`) and `identifier`. And the reason why we use the `identifier` is because of the `a => a + 1` use-case, where you can't have a type for it i.e. `a: Int => a + 1`.

Reply:
okay, so correct me if I'm wrong, but I always thought that these parsers don't actually have the same goal as regular parsers, which have to both
1. reject incorrect code
2. accept correct code

to my eye, tree-sitter parsers actually have different goals, that is
1. accept correct code
2. make predictable ASTs
3. handle bad code gracefully so that the rest of the AST doesn't get ruined

I'm not trying to be pushy, and it's definitely a judgement call as to whether you leave it the way that it is, but in my mind it's preferable that **all `lambda_expression`s have a `bindings` node** than that incorrect lambda expressions are rejected. Of course, if allowing the incorrect `: TypeName` causes conflicts, then I understand. But otherwise, I would actually prefer that ALL `lambda_expressions` have a predictable `bindings` node.

Reply:
It' s a matter of preference, and if you reply "meh I prefer it my way" I'm not gonna push this any further

Reply:
> I always thought that these parsers don't actually have the same goal as regular parsers

Fair point, I was also thinking about it. My main concern is how the AST will be worked upon, once the tree-sitter does it job e.g. when you use nvim-tree-sitter-text-objects you would want your expression to be exactly what they are in order to make a change accordingly. OTOH handling bad code gracefully is something that's definitely required, in this particular case tho it would parse `a: Int => a + 1` as an `ascription_expression`  and an `infix_expression` – so it won't break. I think it's fine to stay as-is :)